Enabling --kokkos-trace option in rocprof-compute

ggottipa-amd opened the pull request:
## Motivation

Adds --kokkos-trace support in the profiling workflow and infrastructure to build and test a mock Kokkos application.

## Technical Details

1)Extracted counter_collection data for kokkos kernels and saved to <workload_dir>/<fbase>_kokkos_kernel_trace.csv
2)CMake configured to build Kokkos and a mock app.
3)Python test infrastructure is updated to profile a mock app with --kokkos-trace enabled. It ensures that the trace files are generated and contain consistent kernel information.

## Test Plan

The test (test_kokkos_trace_output) verifies that all expected trace files are generated and that their contents are consistent and valid for further analysis.

## Test Result

Test test_profile_kokkos passes. Files generated.

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

ggottipa-amd committed:
Enabling --kokkos-trace option

Extracted counter_collection data for kokkos kernels and saved to <workload_dir>/<fbase>_kokkos_kernel_trace.csv

Fixes: SWDEV-520867

ggottipa-amd committed:
Format Fixed.

xuchen-amd commented on projects/rocprofiler-compute/src/utils/utils.py:
```diff
@@ -1099,23 +1099,58 @@ def process_kokkos_trace_output(workload_dir, fbase):
     marker_api_trace_csvs = glob.glob(
```
`marker_api_trace_csvs` is not used, can remove.

xuchen-amd commented on projects/rocprofiler-compute/src/utils/utils.py:
```diff
@@ -1099,23 +1099,58 @@ def process_kokkos_trace_output(workload_dir, fbase):
     )
+    # counter collection csv files are generated for each process
     existing_marker_files_csv = [d for d in marker_api_trace_csvs if path(d).is_file()]
+    existing_counter_files_csv = [
```
the logic suggests the var name `existing_counter_files_csv` is actually `matching_counter_to_marker_files_csv` (or similar).

ggottipa-amd replied:
The number of elements in existing_counter_files_csv can be lower if any counter_collection.csv doesn't exist.

xuchen-amd replied:
~~Please see comment referencing line1106 to 1109.~~

xuchen-amd commented on projects/rocprofiler-compute/src/utils/utils.py:
```diff
@@ -1099,23 +1099,58 @@ def process_kokkos_trace_output(workload_dir, fbase):
+        if path(d).is_file()
+    ]
+    if len(existing_counter_files_csv) != len(existing_marker_files_csv):
+        console_warning(
```
do we expect the number of counter files vs marker files to ALWAYS match?

ggottipa-amd replied:
So far, I see that every process generates a pid_marker_api_trace.csv and pid_counter_collection.csv when run with --kokkos-trace. This is true for the combined results too, at the fbase level. They have a one-to-one correspondence. 
I can remove this check now and add it later if needed (or in ctests).

xuchen-amd reviewed (commented):
Thank you for getting this in!
Just minor comments on the finer details of this new feature.

Overall, looks good! I agree with the approach of joining counter vs marker file to find the correlation given the current limitation.

xuchen-amd commented on projects/rocprofiler-compute/src/utils/utils.py:
```diff
@@ -1099,23 +1099,58 @@ def process_kokkos_trace_output(workload_dir, fbase):
+        for d in [
+            d.replace("marker_api_trace", "counter_collection")
+            for d in existing_marker_files_csv
+        ]
```
~~You are collecting counter collection files based on existing marker_api_trace files, the `existing_counter_files_csv` in your implementation could be a subset of the actual existing counter files since we can't guarantee there is a one-to-one mapping between marker_api_trace files and counter_collection files.~~
Unless this is guaranteed from rocprof?

EDIT:
Just saw you comment, if we verified that there is a one-to-one mapping between the two sets of file, I would suggest to collect counter collection separately (something like `counter_collection = glob.glob(workload_dir + "/out/pmc_1/*/*_counter_collection.csv")`, this would also make your check on line 1112 more meaningful.

ggottipa-amd committed:
Merge remote-tracking branch 'origin' into users/ggottipa-amd/enabling-kokkos-trace

ggottipa-amd committed:
Added test_kokkos_trace_output to ctest

test_kokkos_trace_output runs the profiler on lulesh.kk with --kokkos-trace enabled.

Tests:
1)Verify a match for Kernel_Name and Function,Process_Id , Thread_Id
2)Look for overlap in intervals	(Start_Timestamp,End_Timestamp)
3)Verify the files specific to --kokkos-trace (marker_api_trace and kokkos_kernel_trace in addition to other files).

Additions to rocprof-compute build process:
1) git clones kokkos, kokkos-mini-apps to rocprofiler-compute build directory.
2) builds lulesh app to generate lulesh binary

Fixes: #516740

ggottipa-amd committed:
Cmake Format Fixed. Kokkos build will be skipped if hipcc is not present.

feizheng10 commented on projects/rocprofiler-compute/CMakeLists.txt:
```diff
@@ -41,6 +41,74 @@ set(EXECUTABLE_NAME "rocprof-compute")
+            )
+    endif()
+
+    externalproject_add(
```
Is it FetchContent() better?

ggottipa-amd replied:
FetchContent() works when the packages don't need any changes. But lulesh make file needs changes. We need to find out the architecture for its build.

feizheng10 commented on projects/rocprofiler-compute/CMakeLists.txt:
```diff
@@ -41,6 +41,74 @@ set(EXECUTABLE_NAME "rocprof-compute")
+    # Detect AMD GPU architecture using rocminfo
+    execute_process(
+        COMMAND bash -c
+                "rocminfo | grep -E 'Name: gfx[0-9]+' | head -n 1 | awk '{print \$2}'"
```
It seems not a reliable solution for me. How to handle (a) if cannot find rocminfo, (b) awk won't work for WIN build in the future.

feizheng10 replied:
Any possibility to build Kokkos without specifying target gpu?

ggottipa-amd replied:
No Fei. I tried skipping the gpu info while building lulesh and it didn't work.
So, I am trying to find the underlying architecture differently for each OS. 
I also added fallback for the following,
a) missing rocminfo 
b) missing wmic
c) build failure or missing lulesh binary

feizheng10 commented on projects/rocprofiler-compute/CMakeLists.txt:
```diff
@@ -260,6 +328,13 @@ add_test(
         ${PROJECT_SOURCE_DIR}/tests/test_profile_general.py
     WORKING_DIRECTORY ${PROJECT_SOURCE_DIR})
 
+add_test(
```
Is it better to add if(ENABLE_KOKKOS) here as well?

ggottipa-amd replied:
Yes. I did.

feizheng10 reviewed (changes requested):
And technically, rocprof-compute has no dependency on Kokkos directly.
The only purpose we add it here is for testing.
So, I feel like most these dependency should move into /tests

feizheng10 commented on projects/rocprofiler-compute/CMakeLists.txt:
```diff
@@ -41,6 +41,74 @@ set(EXECUTABLE_NAME "rocprof-compute")
 include(ExternalProject)
 include(GNUInstallDirs)
 
+find_program(HIPCC_EXECUTABLE hipcc)
```
Diverged with tests/CMakeLists.txt amdclang++

ggottipa-amd replied:
True, fixed this one.

ggottipa-amd committed:
Modifying the build process for Kokkos, Lulesh. Building according the underlying OS. Shifting from lulesh make process to cmake.

ggottipa-amd commented:
> And technically, rocprof-compute has no dependency on Kokkos directly. The only purpose we add it here is for testing. So, I feel like most these dependency should move into /tests

Got it. I will make sure that this process is triggered only when tests are enabled and also move the files/folders created for kokkos into /tests.

ggottipa-amd committed:
Ensure Kokkos installation with find_package.
+ other changes to avoid deprecated headers issue. Issue still present.

ggottipa-amd committed:
ENABLE_KOKKOS is set only if ENABLE_TESTS is ON.

ggottipa-amd committed:
CMake format fixed.

ggottipa-amd committed:
Added toggle support for MPI.

ggottipa-amd committed:
CMake format fixed.

ggottipa-amd committed:
Merge remote-tracking branch 'origin' into users/ggottipa-amd/enabling-kokkos-trace

ggottipa-amd committed:
Removed the code that patches lulesh Makefile.

ggottipa-amd committed:
Profiling a mock app to test --kokkos-trace.

ggottipa-amd committed:
CMake format updated.
Mock app size,N, increased.

ggottipa-amd committed:
Merge remote-tracking branch 'origin/develop' into users/ggottipa-amd/enabling-kokkos-trace

ggottipa-amd committed:
Revert "Merge remote-tracking branch 'origin/develop' into users/ggottipa-amd/enabling-kokkos-trace"

This reverts commit 107150b7e26576d307537c34e88c537883bc54bc, reversing
changes made to d33d1c70d778251c073a9d9ab9406199e320cc96.

ggottipa-amd committed:
Updated comments.

ggottipa-amd committed:
Fix format - ruff.

ggottipa-amd committed:
Merge remote-tracking branch 'origin/develop' into users/ggottipa-amd/enabling-kokkos-trace

ggottipa-amd committed:
Fixed issues from rebasing.

ggottipa-amd committed:
Merge remote-tracking branch 'origin/develop' into users/ggottipa-amd/enabling-kokkos-trace

ggottipa-amd committed:
CMake format fixed, gersemi.

ggottipa-amd committed:
Ruff format fix.

ggottipa-amd committed:
Merge remote-tracking branch 'origin/develop' into users/ggottipa-amd/enabling-kokkos-trace

ggottipa-amd committed:
Updated CMAKE_ARGS for kokkos build.

ggottipa-amd committed:
Merge remote-tracking branch 'origin/develop' into users/ggottipa-amd/enabling-kokkos-trace

dayatsin-amd reviewed (changes requested):
Definitely NACK from me.

I only checked the changes in the projects/rocminfo directory and your changes there are incorrect. It seems you are trying to apply changes from an old commit.

dgaliffiAMD commented:
@xuchen-amd , It looks like a merge failure changed the scope of this PR. I suggest closing this one and reopen

vedithal-amd commented:
@ggottipa-amd should we close this against https://github.com/ROCm/rocm-systems/pull/1080?
